### PR TITLE
Final arm64 related functionality

### LINKED ISSUE #757: Necessary functionality support for spawning microVMs on aarch64
- [x] Implement the equivalent of function `arch_memory_regions` for `aarch64` (i.e a function that returns the valid memory addresses for this architecture). Checkout [3.5 Chapter from the A-72 Cortex processor](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.100095_0002_03_en/way1382037547684.html)
and [crosvm] #1003 
(https://chromium.googlesource.com/chromiumos/platform/crosvm/+/master/aarch64/src/lib.rs).
- [x] Platform dependent irq assignment for the mmio device manager #1001.
- [x] Merge https://github.com/nathansizemore/epoll/pull/8.
- [x] Implement specific **aarch64** system configuration (i.e interrupts and registers). See `configure_system` function. Use the the in-kernel GIC support for the hardware VGIC features (additional kernel config  flag required).
- [ ] Create the Flatten Device Tree of the vm that should be spawned.
- [x] Implement specific device allocation for **aarch64** (i.e mmio device allocation)
     *  Is the mmio base address architecture specific?
     * Maybe fix #58?
- [x] Implement kernel loading
- [x] Zero page start for **aarch64**?
- [x] Kernel cmdline start for **aarch64**?
- [x] Working aarch64 kernel config

### PR DESCRIPTION
**Pull request by dianpopa**
With this functionality, I can ssh into an aarch64 guest.
Replication steps to come.

Still some bugs need to be solved:
* time is wrong
* seccomp filters are not applied

Issue #, if available: Fixes #757.

Description of changes:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: vmm/src/lib.rs [hunk 1]
```diff
@@ -2343,14 +2359,15 @@ mod tests {
...
-            host_dev_name: String::from("hostname"),
+            host_dev_name: String::from("hostname1"),
             guest_mac: None,
             rx_rate_limiter: None,
             tx_rate_limiter: None,
             allow_mmds_requests: false,
             tap: None,
         };
-        assert!(vmm.insert_net_device(network_interface).is_ok());
+        vmm.insert_net_device(network_interface).expect("lala");
```
Comment (petreeftime) [inline vmm/src/lib.rs]: The expect message is not very helpful.

File: vmm/src/lib.rs [hunk 2]
```diff
@@ -1342,18 +1337,39 @@ impl Vmm {
...
-        self.attach_virtio_devices()?;
-        self.attach_legacy_devices()?;
+            self.register_events()?;
 
-        let entry_addr = self.load_kernel()?;
+            vcpus = self.create_vcpus(entry_addr, request_ts)?;
+        }
 
-        self.configure_system()?;
+        #[cfg(target_arch = "aarch64")]
```
Comment (petreeftime) [inline vmm/src/lib.rs]: Maybe it's explained in the missing part of the commit message, but it's not clear why these two setup steps are so different. Is it an error on x86_64 to create the vcpus earlier? Why can't the first step be load_kernel on both?

File: vmm/src/lib.rs [hunk 3]
```diff
@@ -1342,18 +1337,39 @@ impl Vmm {
...
-        self.register_events()?;
+            vcpus = self.create_vcpus(entry_addr, request_ts)?;
 
-        let vcpus = self.create_vcpus(entry_addr, request_ts)?;
+            self.setup_interrupt_controller()?;
+
+            self.attach_virtio_devices()?;
+            self.attach_legacy_devices()?;
+
+            self.register_events()?;
```
Comment (petreeftime) [inline vmm/src/lib.rs]: Why are register_events and configure_system swapped?

File: vmm/src/lib.rs [hunk 4]
```diff
@@ -1162,17 +1162,9 @@ impl Vmm {
             .ok_or(StartMicrovmError::VcpusNotConfigured)?;
```
Comment (acatangiu) [inline vmm/src/lib.rs]: Commit `af4d5a0b292dfc6ef6b381d75f76c4f5ff2c3701` has an incomplete description.

File: vmm/src/lib.rs [hunk 5]
```diff
@@ -2343,14 +2359,15 @@ mod tests {
...
+            host_dev_name: String::from("hostname1"),
             guest_mac: None,
             rx_rate_limiter: None,
             tx_rate_limiter: None,
             allow_mmds_requests: false,
             tap: None,
         };
-        assert!(vmm.insert_net_device(network_interface).is_ok());
+        vmm.insert_net_device(network_interface).expect("lala");
+        //assert!(vmm.insert_net_device(network_interface).is_ok());
```
Comment (acatangiu) [inline vmm/src/lib.rs]: Leftover commented line.

File: vmm/src/lib.rs (hunk 3 above)
Comment (dianpopa) [reply vmm/src/lib.rs]: No reason and I also just realized that there is no reason in having them inside the aarch64 cfg. I removed them from there. Take a look.

File: vmm/src/lib.rs (hunk 2 above)
Comment (dianpopa) [reply vmm/src/lib.rs]: I updated the commit message and also put a comment before branching the implementation. Take a look.

Comment (petreeftime) [reply vmm/src/lib.rs]: Looks much cleaner now. 

Could it be possible to further reduce duplication by doing the following:
```
            let entry_addr = self.load_kernel()?;

            #[cfg(target_arch = "aarch64")]
            vcpus = self.create_vcpus(entry_addr, request_ts)?;

            self.setup_interrupt_controller()?;
            self.attach_virtio_devices()?;
            self.attach_legacy_devices()?;

            #[cfg(target_arch = "x86_64")]
            vcpus = self.create_vcpus(entry_addr, request_ts)?;
```

Comment (acatangiu) [reply vmm/src/lib.rs]: nit: Would be nice to also add a comment for both like:
```
// On aarch64 we need to create vcpus before creating the interrupt controller.
#[cfg(target_arch = "aarch64")]
vcpus = self.create_vcpus(entry_addr, request_ts)?;
...
// On x86_64 we need to create vcpus after setting up everything else.
#[cfg(target_arch = "x86_64")]
vcpus = self.create_vcpus(entry_addr, request_ts)?;
```

(Or smth similar)

File: vmm/src/device_manager/mmio.rs [hunk 6]
```diff
@@ -160,18 +171,30 @@ impl MMIODeviceManager {
...
     }
 }
 
+/// Private structure for storing information about the MMIO device registered at some address.
+#[derive(Clone, Debug)]
+pub struct MMIODeviceInfo {
+    addr: u64,
+    irq: u32,
+    len: u64,
+    type_: u32,
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply vmm/src/device_manager/mmio.rs]: > I thought about this but dev_type inside MMIODeviceInfo?

Sounds good to me :laughing:, but like I said, it's a **nit**.

File: vmm/src/device_manager/mmio.rs [hunk 7]
```diff
@@ -160,15 +226,39 @@ impl MMIODeviceManager {
...
+    fn addr(&self) -> u64 {
+        self.addr
+    }
+    fn irq(&self) -> u32 {
+        self.irq
+    }
+    fn length(&self) -> u64 {
+        self.len
+    }
+    fn type_(&self) -> FDTDeviceType {
```
Comment (acatangiu) [inline vmm/src/device_manager/mmio.rs]: What do you think about making `type_` in `MMIODeviceInfo` also an `enum`, ideally the same enum as `FDTDeviceType`.

I think you can call this something generic (not fdt related) and define it for both `x86_64` and `aarch64` in order to use it everywhere.

Comment (dianpopa) [reply vmm/src/device_manager/mmio.rs]: Good suggestion. Take a look.

**Pull request merged**